langchain/src/vectorstores/closevector_web.ts

```diff
@@ -0,0 +1,276 @@
+ */
+
+/**
+ * Arguments for creating a CloseVectorWeb instance, extending CloseVectorHNSWLibArgs.
+ */
+export interface CloseVectorWebArgs
+  extends CloseVectorHNSWLibArgs<HierarchicalNSWT> {
+  instance?: CloseVectorHNSWWeb;
+}
+
+/**
+ * Class that implements a vector store using CloseVector, It extends the SaveableVectorStore class and
+ * provides methods for adding documents and vectors, performing
+ * similarity searches, and saving and loading the vector store.
+ */
+export class CloseVectorWeb extends SaveableVectorStore {
+  declare FilterType: (doc: Document) => boolean;
+
+  _instance?: CloseVectorHNSWWeb;
+
+  // credentials will not be saved to disk
+  credentials?: CloseVectorCredentials;
+
+  _vectorstoreType(): string {
+    return "closevector";
+  }
+
+  constructor(
+    embeddings: Embeddings,
+    args: CloseVectorWebArgs,
+    credentials?: CloseVectorCredentials
+  ) {
+    super(embeddings, args);
+    if (args.instance) {
+      this.instance = args.instance;
+    } else {
+      this.instance = new CloseVectorHNSWWeb(embeddings, args);
+    }
+    this.credentials = credentials;
+  }
+
+  public get instance(): CloseVectorHNSWWeb {
+    if (!this._instance) {
+      throw new Error(
+        "Vector store not initialised yet. Try calling `addTexts` first."
+      );
+    }
+    return this._instance;
+  }
+
+  private set instance(instance: CloseVectorHNSWWeb) {
+    this._instance = instance;
+  }
+
+  /**
+   * Method to add documents to the vector store. It first converts the
+   * documents to vectors using the embeddings, then adds the vectors to the
+   * vector store.
+   * @param documents The documents to be added to the vector store.
+   * @returns A Promise that resolves when the documents have been added.
+   */
+  async addDocuments(documents: Document[]): Promise<void> {
+    await this.instance.addDocuments(documents);
+  }
+
+  /**
+   * Method to save the index to the CloseVector CDN.
+   * @param options
+   * @param options.url the upload url generated by the CloseVector API: https://closevector-docs.getmegaportal.com/docs/api/http-api/file-url
+   * @param options.onProgress a callback function to track the upload progress
+   */
+  async saveToCloud(options: Parameters<typeof this.instance.saveToCloud>[0] & {
+    uuid?: string
+  }) {
+    if (!this.instance.uuid && !options.uuid) {
+      throw new Error("No uuid provided");
+    }
+    if (!this.instance.uuid) {
+      this.instance._uuid = options.uuid;
+    }
+    await this.save(this.instance.uuid);
+    await this.instance.saveToCloud(options);
+  }
+
+  /**
+   * Method to save the vector store to a directory. It saves the HNSW
+   * index, the arguments, and the document store to the directory.
+   * @param directory The directory to which to save the vector store.
+   * @returns A Promise that resolves when the vector store has been saved.
+   */
+  async save(directory: string): Promise<void> {
```

Comment:
What is `directory` in a web context?

Reply:
IndexedDB is used to mock the file system. So, in a web context, a directory is more or less like a key saved in IndexedDB. I have also added comments in the code.